src/legacy/core_plugins/expressions/public/np_ready/public/expression_renderer.tsx

```diff
@@ -17,58 +17,120 @@
  * under the License.
  */
 
-import { useRef, useEffect } from 'react';
+import { useRef, useEffect, useState } from 'react';
 import React from 'react';
-import { ExpressionAST, IExpressionLoaderParams, IInterpreterResult } from './types';
-import { IExpressionLoader, ExpressionLoader } from './loader';
+import classNames from 'classnames';
+import { EuiLoadingChart, EuiProgress } from '@elastic/eui';
+import { ExpressionAST, IExpressionLoaderParams, IInterpreterErrorResult } from './types';
+import { ExpressionLoader } from './loader';
 
 // Accept all options of the runner as props except for the
 // dom element which is provided by the component itself
 export interface ExpressionRendererProps extends IExpressionLoaderParams {
   className: string;
   dataAttrs?: string[];
   expression: string | ExpressionAST;
-  /**
-   * If an element is specified, but the response of the expression run can't be rendered
-   * because it isn't a valid response or the specified renderer isn't available,
-   * this callback is called with the given result.
-   */
-  onRenderFailure?: (result: IInterpreterResult) => void;
+  renderError?: (error?: string | null) => React.ReactElement | React.ReactElement[];
+}
+
+interface State {
+  isEmpty: boolean;
+  isLoading: boolean;
+  error: null | Error;
 }
 
 export type ExpressionRenderer = React.FC<ExpressionRendererProps>;
 
-export const createRenderer = (loader: IExpressionLoader): ExpressionRenderer => ({
+const defaultState: State = {
+  isEmpty: true,
+  isLoading: false,
+  error: null,
+};
+
+export const ExpressionRendererImplementation = ({
   className,
   dataAttrs,
   expression,
-  onRenderFailure,
   ...options
 }: ExpressionRendererProps) => {
   const mountpoint: React.MutableRefObject<null | HTMLDivElement> = useRef(null);
   const handlerRef: React.MutableRefObject<null | ExpressionLoader> = useRef(null);
+  const [state, setState] = useState<State>({ ...defaultState });
+
+  let isMounted = true;
```

Comment:
This will not work, `isMounted` should be a ref or a state variable, using `useState` or `useRef`. You could use [`useMountedState()`](https://github.com/streamich/react-use/blob/master/docs/useMountedState.md), for example.

Reply:
If you just define a variable in render scope like that, it will just create a new `isMounted` variable every time this component re-renders.

Reply:
Fixed in latest by using the existing handlerRef